TypeCobol/Compiler/Diagnostics/CrossChecker.cs

```diff
@@ -590,24 +620,45 @@ public static DataDefinition CheckVariable(Node node, StorageArea storageArea, b
                     }
                 }
 
-                //add the found DataDefinition to a dictionary depending on the storage area type
+                //Initialize node caches for DataDef and Symbol.
+                IDictionary<StorageArea, DataDefinition> dataDefinitionStorage;
+                IDictionary<StorageArea, VariableSymbol> symbolStorage;
                 if (isReadStorageArea)
                 {
-                    //need to initialize the dictionaries
+                    //Initialize reads dictionaries
                     if (node.StorageAreaReadsDataDefinition == null)
-                    {
                         node.StorageAreaReadsDataDefinition = new Dictionary<StorageArea, DataDefinition>();
-                    }
-                    node.StorageAreaReadsDataDefinition.Add(storageArea, dataDefinitionFound);
+                    if (node.StorageAreaReadsSymbol == null)
+                        node.StorageAreaReadsSymbol = new Dictionary<StorageArea, VariableSymbol>();
+
+                    //Target read dictionaries
+                    dataDefinitionStorage = node.StorageAreaReadsDataDefinition;
+                    symbolStorage = node.StorageAreaReadsSymbol;
                 }
                 else
                 {
-                    //need to initialize the dictionaries
+                    //Initialize writes dictionaries
                     if (node.StorageAreaWritesDataDefinition == null)
-                    {
                         node.StorageAreaWritesDataDefinition = new Dictionary<StorageArea, DataDefinition>();
-                    }
-                    node.StorageAreaWritesDataDefinition.Add(storageArea, dataDefinitionFound);
+                    if (node.StorageAreaWritesSymbol == null)
+                        node.StorageAreaWritesSymbol = new Dictionary<StorageArea, VariableSymbol>();
+
+                    //Target writes dictionaries
+                    dataDefinitionStorage = node.StorageAreaWritesDataDefinition;
+                    symbolStorage = node.StorageAreaWritesSymbol;
+                }
+
+                //Add DataDefinition found and corresponding VariableSymbol into caches
+                dataDefinitionStorage.Add(storageArea, dataDefinitionFound);
+                if (dataDefinitionFound.ParentTypeDefinition == null)
```

Comment:
There is one exception here: resolution of `depending on` inside a Typedef need to target a Symbol inside its TypeDefinition. 
But when there is a typed variable inside a Typedef, `depending on` clause can also target an expanded variable.
So it would be:
`if (dataDefinitionFound.ParentTypeDefinition == null || parentTypeDefinition == dataDefinitionFound.ParentTypeDefinition)`

Anyway, I think we simply skip this test. When Type expansion, will be available, the Symbol will be unique.

Reply:
Kept the test for the Debug assertion, so now the dictionary may hold null values (dbd33bfb988818d12e045defdecd7a72f86f72f8).